### ISSUE TITLE
Autowiring & tagging by _defaults does not work together

### ISSUE THREAD

**Issue by gregoirepaqueron** [symfony/monolog-bundle#228]
It looks like the MonologBundle has a priority too high, and runs too early, applying channel tagging before that the services have been autowired. 

In my case : 

```yaml
services:
    _defaults:
        autowire: true
        autoconfigure: true
        public: false
        tags:
            - { name: monolog.logger, channel: technical }

    AppBundle\Manager\:
        resource: '../../../src/AppBundle/Manager'
```

My loggers inside my managers are not handled with `technical` channel.

But if i change the service definitions by : 
```yaml
    AppBundle\Manager\:
        resource: '../../../src/AppBundle/Manager'
        arguments:
            - '@logger'
```
it works.

So it looks like MonologBundle find the tag, but there is no logger inside the service to be overridden. Then, the autowiring will run and will put the default logger inside my services, ignoring the tagging.

**Comment by stof** [comment]
@nicolas-grekas the AutowirePass runs as an optimization pass. This means that the only way to run after it would be to run as an optimization pass at lower priority or before_removing pass.
But this also puts us after the validation passes, as they are also optimization passes at the 0 priority. Do you think it would be acceptable ? I think the only circular reference we could create would be if you try to inject a logger into one of your logging handler, but I'm not sure whether this is worth guarding against or no. What you do think ?

**Comment by nicolas-grekas** [comment]
MonologBundle would pass after AutowirePass and change TypedReference to LoggerInterface by the logger for the channel?
LGTM, no special care required IMHO.

There is another way to configure this manually:
```
_defaults:
  bind:
    $logger: '@logger'
```

**Comment by nicolas-grekas** [comment]
who should MonologBundle know where to inject the channel? AFAIK, it only looks for references to "logger", isn't it?

**Comment by alcaeus** [comment]
FWIW, looking for `logger` is what caused an additional problem when we turned on autowiring, as the reference that was injected wasn't `logger` but rather `monolog.logger`. To fix the issue, we created a new compiler pass which is run as an optimization pass (and thus after `AutowirePass`). This new pass relies on the original `LoggerChannelPass` to create the logger services (these need to be added before optimization) and repeats the code from https://github.com/symfony/monolog-bundle/blob/master/DependencyInjection/Compiler/LoggerChannelPass.php#L37..L65.

I'm not sure if this code could generally be moved to an optimization pass to avoid looping over the services twice; I only know that moving the entire pass to run after `AutowirePass` didn't work in my tests.
